natter-api/src/main/java/com/manning/apisecurityinaction/controller/SpaceController.java

```diff
@@ -50,85 +56,120 @@ public JSONObject createSpace(Request request, Response response) {
   public JSONObject postMessage(Request request, Response response) {
     var spaceId = Long.parseLong(request.params(":spaceId"));
     var json = new JSONObject(request.body());
+
     var user = json.getString("author");
+
     if (!user.matches("[a-zA-Z][a-zA-Z0-9]{0,29}")) {
+
       throw new IllegalArgumentException("invalid username");
+
     }
     var message = json.getString("message");
+
     if (message.length() > 1024) {
+
       throw new IllegalArgumentException("message is too long");
+
     }
 
+
     return database.withTransaction(tx -> {
+
       var msgId = database.findUniqueLong(
           "SELECT NEXT VALUE FOR msg_id_seq;");
+
       database.updateUnique(
+
           "INSERT INTO messages(space_id, msg_id, msg_time," +
               "author, msg_text) " +
               "VALUES(?, ?, current_timestamp, ?, ?)",
+
           spaceId, msgId, user, message);
 
       response.status(201);
+
       var uri = "/spaces/" + spaceId + "/messages/" + msgId;
+
       response.header("Location", uri);
+
       return new JSONObject().put("uri", uri);
+
     });
   }
 
   public Message readMessage(Request request, Response response) {
```

Comment:
Code review for readMessage function:

1. Authorization check is missing
2. spaceId should not be added in url and it can be added if it is encrypted
3. msgId should not be added in url and it can be added if it is encrypted
4. There is no check that the spaceId is owned by logged in user, so hacker can read message from any space
5. There is no check that the msgId is owned by logged in user, so hacker can read message from any space
6. No error handling for missing input/param
7. No error handling for forbidden status
8. Description of function and parameters is not added
9. Remove empty lines